declarative interface for cardano-node

falcucci opened the pull request:
-  [x] Define parameters for specifying a node by network and possible version in the application definitions.
-  [x] Design and implement function to return a node socket based on specified parameters.
-  [x] Reconfigure application to request a node socket with this new method.
-  [x] Test by comparing the received node socket to the desired one.
-  [x] Once successful, eliminate the need to explicitly request a socat container pointing to the right DNS.
-  [x] Thoroughly test new node socket method to ensure stability and accuracy.

> Note: Specifying the version isnt viable ATM since we aren't bootstraping a new cardano-node but connecting to.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a "Cardano Node Connector" trait for Kubernetes applications, enabling enhanced integration with the Cardano blockchain network.

- **Documentation**
  - Updated documentation for the "Cardano Node Connector" trait, detailing its purpose and usage.

- **Infrastructure Updates**
  - Enabled AWS account identity retrieval by updating the `aws_caller_arn` configuration.
  - Added a new Kubernetes TraitDefinition resource for connecting applications to a Cardano Node.
  - Updated configuration parameters to support network and volume mounts for Kubernetes pods.
  - Enabled additional Kubernetes addons and Cardano nodes in the development environment configuration.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

falcucci committed:
add cardano node connector trait definition

adds the definition of `cardano-node-connector` trait definition to the definitions.tf file. the trait definition provides an application with the ability to connect to a cardano node.

falcucci committed:
minor changes

shlevy commented on infra/modules/kubevela-addons/definitions.tf:
```diff
@@ -162,6 +162,33 @@ resource "kubectl_manifest" "componentdefinition_bucket" {
 }
 
+resource "kubectl_manifest" "traitdefinition_cardano_node_connector" {
+  force_new = true
```
Out of curiosity, why is this needed? Do `TraitDefinitions` not support in-place modification?

falcucci replied:
you are right @shlevy 

removed :)

shlevy commented on infra/modules/kubevela-addons/definitions/cardano-node-connector.cue:
```diff
@@ -0,0 +1,34 @@
+parameter: {
+  network: *"preview" | string
+  version?: *"latest" | string
+  volumeMounts: *[{
```
I see this is still a draft, is the plan to generate the volumeMount stanza from a requested mount location?

falcucci replied:
yep, the intention was to make it as dynamical as possible.

some applications could use a different `node.sock` path instead of just /ipc or maybe even in the case they use their cardano-node instance in a different default path.

we take /ipc as default tho and let the user override those values if needed,

sounds reasonable?

shlevy replied:
Yes, thanks!

falcucci committed:
removed `force_new = true` from the `kubectl_manifest` resource for `traitdefinition_cardano_node_connector`.

falcucci committed:
updates the `kubectl_manifest` to use a `file` resource instead of `templatefile`

the `infra/account.hcl` file now includes an `aws_caller_arn` local to get the aws caller identity arn. this ensures that the resource is properly configured and the identity of the caller is correctly identified.

the `kubectl_manifest` resource in `infra/modules/kubevela-addons/definitions.tf` now uses a `file` resource instead of a `templatefile` resource. this ensures that the resource is properly configured and the file is properly read.

falcucci committed:
docs: add new `cardano/node-connector.md` placeholder file

- Add a new file `cardano/node-connector.md` with a placeholder content

Signed-off-by: Alexsander Falcucci <alex.falcucci@gmail.com>

falcucci committed:
feat: add new application "oura" with cardano integration

- add a new file `doc/kubevela-definitions/cardano/example.yaml`
- define an application named `oura` in the `oura` namespace
- create a component named `oura` of type `webservice` with an image `alexfalcucci/oura:2.0.1`
- add a trait of type `cardano-node-connector` to the `oura` component, with the property `network` set to `preview`
- define two policies named `debug` and `local-vela-system`
- add a property `clusters` to the `local-vela-system` policy, with a value of `local`
- add a property `namespace` to the `local-vela-system` policy, with a value of `vela-system`
- define a workflow with mode `steps` set to `dag`
- add a step named `local-vela-system` to the workflow with an alias `deploy to local-vela-system`
- add a property `policies` to the `local-vela-system` step, with a value of `local-vela-system`
- define a topology type with the name `local-vela-system` and a property `clusters` with a value of `local` and `namespace` with a value of `vela-system`
- deploy the `local-vela-system` step with the `local-vela-system

falcucci committed:
chore: remove workflow section from example.yaml file

- removed the `workflow` section from the `example.yaml` file.

falcucci committed:
chore: enable addons and cardano nodes in terragrunt config

- switch `enable_addons` and `enable_cardano_nodes` from `false` to `true` in the terragrunt configuration file.

falcucci committed:
removed `debug` and `local-vela-system` policies from the cardano example spec

- removed the specification of local clusters and namespaces under `local-vela-system` policy

falcucci committed:
refactor: cardano-node-connector for cleaner structure

- removed the import of `encoding/json` in `cardano-node-connector.cue`
- removed the `version` parameter in `cardano-node-connector.cue`

falcucci committed:
docs: improve cardano node connector trait documentation in kubvela

- add documentation for the cardano node connector trait in kubvela
- provide an overview, including optional parameters and volume-specific configurations for the connector trait
- provide a usage example based on an instance of a kubvela application.

falcucci committed:
merge 'main'

renebarbosafl commented:
Input:

```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: chain-indexer-mainnet-demo
  namespace: marlowe-staging
spec:
  components:
  - name: chain-indexer-mainnet-demo
    properties:
      cpu: "1"
      env:
      - name: NODE_CONFIG
        value: /node-config/network/mainnet/cardano-node/config.json
      - name: CARDANO_NODE_SOCKET_PATH
        value: /ipc/node.socket
      - name: HTTP_PORT
        value: "3781"
      - name: OTEL_EXPORTER_OTLP_ENDPOINT
        value: http://k8s-monitoring-grafana-agent.grafana-agent:4318
      - name: OTEL_SERVICE_NAME
        value: marlowe-chain-indexer-mainnet-demo
      image: ghcr.io/input-output-hk/marlowe-chain-indexer:0.0.5
      imagePullPolicy: Always
      memory: 8Gi
    traits:
    - properties:
        appMountPath: /node-config
        args:
        - clone
        - --single-branch
        - --
        - https://github.com/input-output-hk/cardano-configurations
        - /node-config
        image: alpine/git
        initMountPath: /node-config
        mountName: node-config
        name: node-config-cloner
      type: init-container
    - properties:
        network: mainnnet
      type: cardano-node-connector
    - properties:
        name: chain-mainnet-demo
        allocateDatabase:
          storage: 400
          instanceClass: db.t3.medium
      type: postgres-instance
    type: webservice
```

renebarbosafl commented:
Outputs:

```yaml
---
# Application(chain-indexer-mainnet-demo) -- Component(chain-indexer-mainnet-demo)
---

apiVersion: apps/v1
kind: Deployment
metadata:
  annotations: {}
  labels:
    app.oam.dev/appRevision: ""
    app.oam.dev/component: chain-indexer-mainnet-demo
    app.oam.dev/name: chain-indexer-mainnet-demo
    app.oam.dev/namespace: crossplane
    app.oam.dev/resourceType: WORKLOAD
    workload.oam.dev/type: webservice
  name: chain-indexer-mainnet-demo
  namespace: crossplane
spec:
  selector:
    matchLabels:
      app.oam.dev/component: chain-indexer-mainnet-demo
  template:
    metadata:
      labels:
        app.oam.dev/component: chain-indexer-mainnet-demo
        app.oam.dev/name: chain-indexer-mainnet-demo
    spec:
      containers:
      - env:
        - name: NODE_CONFIG
          value: /node-config/network/mainnet/cardano-node/config.json
        - name: CARDANO_NODE_SOCKET_PATH
          value: /ipc/node.socket
        - name: HTTP_PORT
          value: "3781"
        - name: OTEL_EXPORTER_OTLP_ENDPOINT
          value: http://k8s-monitoring-grafana-agent.grafana-agent:4318
        - name: OTEL_SERVICE_NAME
          value: marlowe-chain-indexer-mainnet-demo
        - name: DB_USER
          valueFrom:
            secretKeyRef:
              key: username
              name: conn-chainsync-mainnet-demo
        - name: DB_PASS
          valueFrom:
            secretKeyRef:
              key: password
              name: conn-chainsync-mainnet-demo
        - name: DB_HOST
          valueFrom:
            secretKeyRef:
              key: endpoint
              name: conn-chainsync-mainnet-demo
        - name: DB_NAME
          value: scdedb
        image: ghcr.io/input-output-hk/marlowe-chain-indexer:0.0.5
        imagePullPolicy: Always
        name: chain-indexer-mainnet-demo
        resources:
          limits:
            cpu: "1"
            memory: 8Gi
          requests:
            cpu: "1"
            memory: 8Gi
        volumeMounts:
        - mountPath: /node-config
          name: node-config
      - args:
        - UNIX-LISTEN:/ipc/node.socket,fork
        - TCP-CONNECT:cardano-node-mainnnet.vela-system:8090
        image: alpine/socat
        name: socat
        volumeMounts:
        - mountPath: /ipc
          name: ipc
      initContainers:
      - args:
        - clone
        - --single-branch
        - --
        - https://github.com/input-output-hk/cardano-configurations
        - /node-config
        image: alpine/git
        imagePullPolicy: IfNotPresent
        name: node-config-cloner
        volumeMounts:
        - mountPath: /node-config
          name: node-config
      volumes:
      - emptyDir: {}
        name: node-config

---
## From the trait postgres-instance
apiVersion: rds.aws.crossplane.io/v1alpha1
kind: DBParameterGroup
metadata:
  annotations: {}
  labels:
    app.oam.dev/appRevision: ""
    app.oam.dev/component: chain-indexer-mainnet-demo
    app.oam.dev/name: chain-indexer-mainnet-demo
    app.oam.dev/namespace: crossplane
    app.oam.dev/resourceType: TRAIT
    trait.oam.dev/resource: db_parameter_group
    trait.oam.dev/type: postgres-instance
  name: dbpg-chainsync-mainnet-demo
  namespace: crossplane
spec:
  forProvider:
    dbParameterGroupFamilySelector:
      engine: postgres
    description: Parameter group for chainsync-mainnet-demo RDS Instance
    region: us-east-1
    tags:
    - key: Project
      value: scde
    - key: Tribe
      value: smartcontracts
    - key: Organization
      value: iog
    - key: managedBy
      value: kubeVela
    - key: Application
      value: chain-indexer-mainnet-demo
  providerConfigRef:
    name: aws-provider

---
## From the trait postgres-instance
apiVersion: database.aws.crossplane.io/v1beta1
kind: RDSInstance
metadata:
  annotations: {}
  labels:
    app.oam.dev/appRevision: ""
    app.oam.dev/component: chain-indexer-mainnet-demo
    app.oam.dev/name: chain-indexer-mainnet-demo
    app.oam.dev/namespace: crossplane
    app.oam.dev/resourceType: TRAIT
    trait.oam.dev/resource: db_instance
    trait.oam.dev/type: postgres-instance
  name: chain-mainnet-demo
  namespace: crossplane
spec:
  forProvider:
    allocatedStorage: 400
    applyModificationsImmediately: true
    autoMinorVersionUpgrade: false
    caCertificateIdentifier: rds-ca-rsa2048-g1
    copyTagsToSnapshot: false
    dbInstanceClass: db.t3.medium
    dbName: scdedb
    dbParameterGroupName: dbpg-chain-mainnet-demo
    dbSubnetGroupName: dev-subnet-group
    deletionProtection: false
    enableIAMDatabaseAuthentication: false
    enablePerformanceInsights: true
    engine: postgres
    masterUsername: postgres
    maxAllocatedStorage: 1200
    monitoringInterval: 60
    monitoringRoleArn: arn:aws:iam::677160962006:role/rds-monitoring-role
    multiAZ: true
    publiclyAccessible: false
    region: us-east-1
    skipFinalSnapshotBeforeDeletion: true
    storageEncrypted: true
    storageType: gp3
    tags:
    - key: Project
      value: scde
    - key: Tribe
      value: smartcontracts
    - key: Organization
      value: iog
    - key: managedBy
      value: kubeVela
    - key: Application
      value: chain-mainnet-demo
    vpcSecurityGroupIds:
    - sg-0a5910de63f172835
  providerConfigRef:
    name: aws-provider
  writeConnectionSecretToRef:
    name: conn-chain-mainnet-demo
    namespace: crossplane

---
```

renebarbosafl reviewed (changes requested):
Missing `volumes` and `volumeMounts` in the dry-run outputs.

falcucci commented:
> Missing `volumes` and `volumeMounts` in the dry-run outputs.

@renebarbosafl thanks, gonna replicate this app as soon as I get back home

renebarbosafl committed:
adding init-container and fix volume patch

renebarbosafl committed:
updating doc

shlevy commented on doc/kubevela-definitions/cardano/node-connector.md:
```diff
@@ -0,0 +1,109 @@
+
+The Cardano Node Connector in Kubvela acts as a bridge between your application and the Cardano node through specific application characteristics and settings
+
+The ***required*** parameter for implementing this trait is:
```
How is it required if it has a default?

renebarbosafl replied:
I believe it's just a terminology confusion. The "default" value here is just for VelaUX , when someone is trying to configure the trait using the UI, "preview" is the first value available on the switchBox. So yeah, I think we have to update this doc to avoid such confusion.

shlevy commented on doc/kubevela-definitions/cardano/node-connector.md:
```diff
@@ -0,0 +1,109 @@
+ - `network`: Defines the network choice for the Cardano Node, with a default setting as "preview". 
+<br>The other options available are "preprod" and "mainnet".
+
+This configuration effectively attaches specific volumes to each container that is part of your application. Also, a `socat` container bridges your application and the Cardano Node by utilizing TCP Socket over a UNIX Socket.
```
This is an implementation detail which may change, all the user needs to know is that there will be a volume with the node socket.

shlevy commented on doc/kubevela-definitions/cardano/node-connector.md:
````diff
@@ -0,0 +1,109 @@
+        network: preview
+```
+
+This is the output from the Application code above:
````
This is not useful for end-user documentation. Instead, we should document where the socket can be found once you add your trait.
